distribution: add back resumable pulls

jakogut opened the pull request:
Upstream docker supports resuming pulls after transient failures using
HTTP range requests. In commit 565fc34, this was replaced with streaming
the tarball directly to disk, obviating the need for a temporary file,
and reducing the amount of free space required to download images.

Add back the upstream version with resumable pulls, but default to the
current behavior of streaming directly to disk. The upstream behavior
can be specified with the environment variable:

BALENA_ENGINE_PULL_METHOD=resumable

Change-type: patch
Signed-off-by: Joseph Kogut <joseph@balena.io>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/balena-os/balena-engine/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For general information on balenaEngine visit https://www.balena.io/engine

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

robertgzr commented on distribution/pull_v2.go:
```diff
@@ -218,29 +379,59 @@ func (ld *v2LayerDescriptor) DeltaBase() io.ReadSeeker {
+}
+
+func (ld *v2LayerDescriptor) Download(ctx context.Context, progressOutput progress.Output) (io.ReadCloser, int64, error) {
+	if os.Getenv("BALENA_ENGINE_PULL_METHOD") == "resumable" {
```
we might want to use `BALENA_PULL_METHOD` to keep the balena specific env vars uniform across the code base, I used the simple `BALENA_*` prefix for the storage migration code...

jakogut commented:
@robertgzr I've gone ahead and changed `BALENA_ENGINE_PULL_METHOD` to `BALENA_PULL_METHOD` as suggested. I've also changed the `stream-to-disk` value to simply `streaming`, and clarified some of the comments a bit more.

jakogut commented:
Also just updated the seek flags to match upstream, using constants from `io` instead of `os`. For example, `io.SeekStart` instead of `os.SEEK_SET`.

jakogut commented:
There's also a test here as well: https://github.com/docker-archive/golem/pull/6/files

robertgzr commented:
@jakogut in yesterdays arch call we concluded it's probably worth discarding our implementation altogether and to just go with the upstream behavior

jakogut commented:
@balena-ci rebase

jakogut commented:
@balena-ci rebase

20k-ultra commented:
@robertgzr @jakogut what's the plan with this PR ? How do we use with the upstream behavior for this ?

lmbarros commented:
@20k-ultra I intend to test this PR today. If it is OK, I believe we shall have a call to discuss if we want to go ahead and merge it. This would be a temporary solution for some issues with large images we have been seeing. (I am looking into a longer-term solution).

jakogut commented:
For context, I changed this to a draft after removing the streaming pull method, as discussed in the arch call, until this can be tested again.

lmbarros commented:
Alternative PR to this one: https://github.com/balena-os/balena-engine/pull/253
